Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you

1- to add enough Markdown cells to explain what you decided to do and to report any observations you make.¶

2-to generate maps to visualize your neighborhoods and how they cluster together

In [1]:
!conda install -c conda-forge beautifulsoup4 --yes

!conda install -c conda-forge geopy --yes

!conda install -c conda-forge folium=0.5.0 --yes

print('Libraries installed!')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries installed!


In [6]:
import numpy as np
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import requests
from pandas.io.json import json_normalize
import json

import requests

from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim

import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

from sklearn.cluster import KMeans

print('Libraries imported!')

Libraries imported!


# Part 1: Web scraping for Toronto neighborhood and build a clean dataframe¶

Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.

The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.
If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.
Note: There are different website scraping libraries and packages in Python. One of the most common packages is BeautifulSoup. Here is the package's main documentation page: http://beautiful-soup-4.readthedocs.io/en/latest/

In [3]:

In [7]:
# open Wiki page with Beautiful Soup
data = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(data, 'html.parser')

Scrape the HTML
The data we want is in a table, with 3 columns PostalCode, Borough and Neighborhood.
The table contains a list of postal codes in Canada where the first letter is M. Postal codes beginning with M are located within the city of Toronto in the province of Ontario.
Only the first three characters are listed, corresponding to the Forward Sortation Area.

In [8]:
postalCodeList = []
boroughList = []
neighborhoodList = []

for row in soup.find('table').find_all('tr'):
    cells = row.find_all('td')
    if(len(cells) > 0):
        postalCodeList.append(cells[0].text)
        boroughList.append(cells[1].text)
        neighborhoodList.append(cells[2].text.rstrip('\n')) # remove the new line char from neighborhood cell

Put the scraped data into dataframe
After scraping the html, we get 3 lists containing the PostalCodes, Boroughs and Neighborhoods in order as in Wiki page.
We will put those 3 lists into a dataframe with corresponding columns name.

Put the scraped data into dataframe
After scraping the html, we get 3 lists containing the PostalCodes, Boroughs and Neighborhoods in order as in Wiki page.
We will put those 3 lists into a dataframe with corresponding columns name.

In [10]:
toronto_neighorhood = [('PostalCode', postalCodeList),
                      ('Borough', boroughList),
                      ('Neighborhood', neighborhoodList)]
toronto_df = pd.DataFrame.from_dict(dict(toronto_neighorhood))
toronto_df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [11]:
toronto_df_dropna = toronto_df[toronto_df.Borough != 'Not assigned'].reset_index(drop=True)
toronto_df_dropna.head()

PostalCode           Borough      Neighborhood
0        M3A        North York         Parkwoods
1        M4A        North York  Victoria Village
2        M5A  Downtown Toronto      Harbourfront
3        M6A        North York  Lawrence Heights
4        M6A        North York    Lawrence Manor

Group neighborhoods by postal and borough
There are some neighborhoods that belongs to the same postal code and borough (Ex: M5A Downtown Toronto has 2 neighborhoods Harbourfront and Regent Park).
We will concat them into same row, seperated by a colon.

In [12]:
toronto_df_grouped = toronto_df_dropna.groupby(['PostalCode','Borough'], as_index=False).agg(lambda x: ','.join(x))
toronto_df_grouped.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Rouge,Malvern
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

Deal with Not assigned Neighborhood
For M7A Queen's Park, there is no neighborhood assigned.
We will replace the 'Not assigned' with the value of the corresponding Borough

In [13]:
na_neigh_rows = toronto_df_grouped.Neighborhood == 'Not assigned'
toronto_df_grouped.loc[na_neigh_rows, 'Neighborhood'] = toronto_df_grouped.loc[na_neigh_rows, 'Borough']
toronto_df_grouped[na_neigh_rows]

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

In [14]:
toronto_df_cleaned = toronto_df_grouped
toronto_df_cleaned.shape

(103, 3)


Part 2: Getting coordinates and add to the Toronto DataFrame

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates.

geolocator = Nominatim(user_agent="tl-toronto-neigh")

postalList = toronto_df_cleaned['PostalCode'].values latList = [] longList = []

for post in postalList: location = None while(location is None): location = geolocator.geocode('{}, Toronto, Ontario'.format(post)) if(location != None): lat = location.latitude long = location.longitude print(post, lat, long) latList.append(lat) longList.append(long)

toronto_coors = [('Postal Code', postalList), ('Latitude', latList), ('Longitude', longList)] coors = pd.DataFrame.from_items(toronto_coors)

Note: Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

The csv file format has:

3 columns: Postal Code, Latitude and Longitude
103 rows: corresponding to 103 postal codes in our toronto dataframe

In [16]:
!wget -q -O "toronto_coordinates.csv" http://cocl.us/Geospatial_data
print('Coordinates downloaded!')
coors = pd.read_csv('toronto_coordinates.csv')

Coordinates downloaded!


In [17]:
print(coors.shape)
coors.head()

(103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [18]:
toronto_df_temp = toronto_df_cleaned.set_index('PostalCode')
coors_temp = coors.set_index('Postal Code')
toronto_df_coors = pd.concat([toronto_df_temp, coors_temp], axis=1, join='inner')

In [19]:
toronto_df_coors.index.name = 'PostalCode'
toronto_df_coors.reset_index(inplace=True)

Check the toronto dataframe with coordinates

In [20]:
print(toronto_df_coors.shape)
toronto_df_coors.head()

(103, 5)


PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

# Part 3: Explore and cluster the neighborhoods in Toronto

In [21]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


Using Folium to create a Map of Toronto with Boroughs markers on top

In [22]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df_coors['Latitude'], toronto_df_coors['Longitude'], toronto_df_coors['PostalCode'], toronto_df_coors['Borough'], toronto_df_coors['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Reduce the number of Boroughs to explore To reduce the numbers of calls to FourSquare API, we will only explore boroughs that have Toronto in their names.

In [23]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df_coors[toronto_df_coors['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(39, 5)


PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [24]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

In [25]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [26]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1714, 9)


PostalCode       Borough                 Neighborhood  BoroughLatitude  \
0        M4E  East Toronto                  The Beaches        43.676357   
1        M4E  East Toronto                  The Beaches        43.676357   
2        M4E  East Toronto                  The Beaches        43.676357   
3        M4E  East Toronto                  The Beaches        43.676357   
4        M4K  East Toronto  The Danforth West,Riverdale        43.679557   

   BoroughLongitude                           VenueName  VenueLatitude  \
0        -79.293031                   Glen Manor Ravine      43.676821   
1        -79.293031  The Big Carrot Natural Food Market      43.678879   
2        -79.293031                 Grover Pub and Grub      43.679181   
3        -79.293031                       Upper Beaches      43.680563   
4        -79.352188                            Pantheon      43.677621   

   VenueLongitude      VenueCategory  
0      -79.293942              Trail  
1      -79.297734  Health Food Store  
2      -79.297215                Pub  
3      -79.292869       Neighborhood  
4      -79.351434   Greek Restaurant

Check how many venues returned for each postal code

In [27]:
venues_df.groupby(['PostalCode', 'Borough', 'Neighborhood'])['VenueName'].count()

PostalCode  Borough           Neighborhood                                                                                        
M4E         East Toronto      The Beaches                                                                                               4
M4K         East Toronto      The Danforth West,Riverdale                                                                              41
M4L         East Toronto      The Beaches West,India Bazaar                                                                            22
M4M         East Toronto      Studio District                                                                                          42
M4N         Central Toronto   Lawrence Park                                                                                             3
M4P         Central Toronto   Davisville North                                                                                          8
M4R         Central Toronto   North Toron

Check how many kinds of venues categories is returned

In [28]:
len(venues_df['VenueCategory'].unique())

233

Analyze venues in each area

In [29]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1714, 236)


PostalCode       Borough                Neighborhoods  Afghan Restaurant  \
0        M4E  East Toronto                  The Beaches                  0   
1        M4E  East Toronto                  The Beaches                  0   
2        M4E  East Toronto                  The Beaches                  0   
3        M4E  East Toronto                  The Beaches                  0   
4        M4K  East Toronto  The Danforth West,Riverdale                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0                   0       0           0              0          0         0   
1                   0       0           0              0          0         0   
2                   0       0           0              0          0         0   
3                   0       0           0              0          0         0   
4                   0       0           0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0

Get the frequency of occurance of each category in an area

In [30]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['PostalCode', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(39, 236)


PostalCode          Borough                  Neighborhoods  \
0        M4E     East Toronto                    The Beaches   
1        M4K     East Toronto    The Danforth West,Riverdale   
2        M4L     East Toronto  The Beaches West,India Bazaar   
3        M4M     East Toronto                Studio District   
4        M4N  Central Toronto                  Lawrence Park   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.024390   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.047619   
4             0.0              0.0               0.0             0.000000   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0           0.0       0.0          0.0                  0.0               0.0   
1           0.0       0.0          0.0                  0.0               0.0   
2           0.0       0.0          0.0                  0.0               0.0   
3           0.0       0.0          0.0                  0.0               0.0   
4           0.0       0.0          0.0                  0.0               0.0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0            0.0        0.0         0.0         0.0   
1                 0.0            0.0        0.0         0.0         0.0   
2                 0.0            0.0        0.0         0.0         0.0   
3                 0.0            0.0        0.0         0.0         0.0   
4                 0.0            0.0        0.0         0.0         0.0   

     Bakery     Bank      Bar  Baseball Stadium  Basketball Stadium  Beach  \
0  0.000000  0.00000  0.00000               0.0                 0.0    0.0   
1  0.024390  0.00000  0.00000               0.0                 0.0    0.0   
2  0.000000  0.00000  0.00000               0.0                 0.0    0.0   
3  0.047619  0.02381  0.02381               0.0                 0.0    0.0   
4  0.000000  0.00000  0.00000               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0              0.0       0.0         0.0                 0.0     0.0   
1              0.0       0.0         0.0                 0.0     0.0   
2              0.0       0.0         0.0                 0.0     0.0   
3              0.0       0.0         0.0                 0.0     0.0   
4              0.0       0.0         0.0                 0.0     0.0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0    0.000000            0.0    0.00000       0.0                   0.0   
1    0.000000            0.0    0.02439       0.0                   0.0   
2    0.045455            0.0    0.00000       0.0                   0.0   
3    0.000000            0.0    0.02381       0.0                   0.0   
4    0.000000            0.0    0.00000       0.0                   0.0   

   Breakfast Spot   Brewery  Bubble Tea Shop  Building  Burger Joint  \
0             0.0  0.000000          0.00000       0.0           0.0   
1             0.0  0.024390          0.02439       0.0           0.0   
2             0.0  0.045455          0.00000       0.0           0.0   
3             0.0  0.047619          0.00000       0.0           0.0   
4             0.0  0.000000          0.00000       0.0           0.0   

   Burrito Place  Bus Line 

Get 10 most occurance venue types in each area

In [31]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['PostalCode'] = toronto_central_venues_freq['PostalCode']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
26        M5T  Downtown Toronto   
38        M7Y      East Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
36        M6S      West Toronto   
25        M5S  Downtown Toronto   
19        M5J  Downtown Toronto   
18        M5H  Downtown Toronto   
16        M5E  Downtown Toronto   
21        M5L  Downtown Toronto   
20        M5K  Downtown Toronto   
15        M5C  Downtown Toronto   
29        M5X  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
11        M4X  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
37        M7A  Downtown Toronto   
28        M5W  Downtown Toronto   
7         M4S   Central Toronto   
22        M5N   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
23        M5P   Central Toronto   
34        M6P      West Toronto   
10        M4W  Downtown Toronto   
2         M4L      East Toronto   
4         M4N   Central Toronto   
31        M6H      West Toronto   
9         M4V   Central Toronto   
8         M4T   Central Toronto   
24        M5R   Central Toronto   
0         M4E      East Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   
32                            Little Portugal,Trinity                   Bar   
26            Chinatown,Grange Park,Kensington Market                   Bar   
38  Business Reply Mail Processing Centre 969 Eastern               Brewery   
33         Brockton,Exhibition Place,Parkdale Village                  Café   
3                                     Studio District                  Café   
36                                  Runnymede,Swansea                  Café   
25                      Harbord,University of Toronto                  Café   
19    Harbourfront East,Toronto Islands,Union Station           Coffee Shop   
18                             Adelaide,King,Richmond           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
21                      Commerce Court,Victoria Hotel           Coffee Shop   
20            Design Exchange,Toronto Dominion Centre           Coffee Shop   
15                                     St. James Town           Coffee Shop   
29              First Canadian Place,Underground city           Coffee Shop   
14                            Ryerson,Garden District           Coffee Shop   
6                                  North Toronto West           Coffee Shop   
11                         Cabbagetown,St. James Town           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                                       Harbourfront           Coffee Shop   
37                                       Queen's Park           Coffee Shop   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
7                                          Davisville          Dessert Shop   
22                                           Roselawn                Garden   
35                              Parkdale,Roncesvalles             Gift Shop   
1                         The Danforth West,Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
5                                    Davisville North                   Gym   
23                 Forest Hill North,Forest Hill West         Jewelry Store   
34                       High Park,The Junction South    Mexican Restaurant   
10                                           Rosedale                  Par

# Clustering areas
Use KMeans algorigthm, try to cluster the toronto central areas into 3 clusters

In [32]:
kclusters = 3

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['PostalCode', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df
toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('PostalCode'), on='PostalCode')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df

PostalCode           Borough  \
27        M5V  Downtown Toronto   
32        M6J      West Toronto   
26        M5T  Downtown Toronto   
38        M7Y      East Toronto   
33        M6K      West Toronto   
3         M4M      East Toronto   
36        M6S      West Toronto   
25        M5S  Downtown Toronto   
19        M5J  Downtown Toronto   
18        M5H  Downtown Toronto   
16        M5E  Downtown Toronto   
21        M5L  Downtown Toronto   
20        M5K  Downtown Toronto   
15        M5C  Downtown Toronto   
29        M5X  Downtown Toronto   
14        M5B  Downtown Toronto   
6         M4R   Central Toronto   
11        M4X  Downtown Toronto   
17        M5G  Downtown Toronto   
12        M4Y  Downtown Toronto   
13        M5A  Downtown Toronto   
37        M7A  Downtown Toronto   
28        M5W  Downtown Toronto   
7         M4S   Central Toronto   
35        M6R      West Toronto   
1         M4K      East Toronto   
30        M6G  Downtown Toronto   
5         M4P   Central Toronto   
23        M5P   Central Toronto   
34        M6P      West Toronto   
10        M4W  Downtown Toronto   
2         M4L      East Toronto   
4         M4N   Central Toronto   
31        M6H      West Toronto   
9         M4V   Central Toronto   
24        M5R   Central Toronto   
0         M4E      East Toronto   
8         M4T   Central Toronto   
22        M5N   Central Toronto   

                                         Neighborhood   Latitude  Longitude  \
27  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420   
32                            Little Portugal,Trinity  43.647927 -79.419750   
26            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049   
38  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   
33         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191   
3                                     Studio District  43.659526 -79.340923   
36                                  Runnymede,Swansea  43.651571 -79.484450   
25                      Harbord,University of Toronto  43.662696 -79.400049   
19    Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752   
18                             Adelaide,King,Richmond  43.650571 -79.384568   
16                                        Berczy Park  43.644771 -79.373306   
21                      Commerce Court,Victoria Hotel  43.648198 -79.379817   
20            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576   
15                                     St. James Town  43.651494 -79.375418   
29              First Canadian Place,Underground city  43.648429 -79.382280   
14                            Ryerson,Garden District  43.657162 -79.378937   
6                                  North Toronto West  43.715383 -79.405678   
11                         Cabbagetown,St. James Town  43.667967 -79.367675   
17                                 Central Bay Street  43.657952 -79.387383   
12                               Church and Wellesley  43.665860 -79.383160   
13                                       Harbourfront  43.654260 -79.360636   
37                                       Queen's Park  43.662301 -79.389494   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
7                                          Davisville  43.704324 -79.388790   
35                              Parkdale,Roncesvalles  43.648960 -79.456325   
1                         The Danforth West,Riverdale  43.679557 -79.352188   
30                                           Christie  43.669542 -79.422564   
5                                    Davisville North  43.712751 -79.390197   
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307   
34                       High Park,The Junction South  43.661608 -79.464763   
10                                           Rosedale  43.679563 -79.377529   
2                       The Beaches West,India Bazaar  43.668999 -79.31557

# Show those clusters onto a map

In [33]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['PostalCode'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Upon observing the result, we can name the clusters as follow:

Cluster 0: Living area (with mostly park, trail, school; and some small businesses)
Cluster 1: Roselawn - Central Toronto (nothing here except a garden)
Cluster 2: Business area (with lots of business venues)